dbms/src/Storages/DeltaMerge/Filter/IntegerSet.cpp

```diff
@@ -0,0 +1,609 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Storages/DeltaMerge/Filter/IntegerSet.h>
+#include <Storages/DeltaMerge/Index/InvertedIndex/Reader.h>
+
+#include <algorithm>
+#include <limits>
+#include <vector>
+
+
+namespace DB::DM
+{
+
+IntegerSetPtr IntegerSet::createValueSet(TypeIndex type_index, const Fields & values)
+{
+    switch (type_index)
+    {
+    case TypeIndex::UInt8:
+        return std::make_shared<ValueSet<UInt8>>(values);
+    case TypeIndex::UInt16:
+        return std::make_shared<ValueSet<UInt16>>(values);
+    case TypeIndex::UInt32:
+        return std::make_shared<ValueSet<UInt32>>(values);
+    case TypeIndex::UInt64:
+        return std::make_shared<ValueSet<UInt64>>(values);
+    case TypeIndex::Int8:
+        return std::make_shared<ValueSet<Int8>>(values);
+    case TypeIndex::Int16:
+        return std::make_shared<ValueSet<Int16>>(values);
+    case TypeIndex::Int32:
+        return std::make_shared<ValueSet<Int32>>(values);
+    case TypeIndex::Int64:
+        return std::make_shared<ValueSet<Int64>>(values);
+    case TypeIndex::Date:
+        return std::make_shared<ValueSet<UInt16>>(values);
+    case TypeIndex::DateTime:
+        return std::make_shared<ValueSet<UInt32>>(values);
+    case TypeIndex::Enum8:
+        return std::make_shared<ValueSet<Int8>>(values);
+    case TypeIndex::Enum16:
+        return std::make_shared<ValueSet<Int16>>(values);
+    case TypeIndex::MyDate:
+    case TypeIndex::MyDateTime:
+    case TypeIndex::MyTimeStamp:
+        return std::make_shared<ValueSet<UInt64>>(values);
+    case TypeIndex::MyTime:
+        return std::make_shared<ValueSet<Int64>>(values);
+    default:
+        return nullptr;
+    }
+}
+
+IntegerSetPtr IntegerSet::createLessRangeSet(TypeIndex type_index, Field max, bool not_included)
+{
+    switch (type_index)
+    {
+    case TypeIndex::UInt8:
+        return std::make_shared<RangeSet<UInt8>>(0, max.get<UInt8>() - not_included);
```

Comment:
Could we run into a case that `max.get<UInt8> == 0`?

Reply:
Actually tidb optimizer can handle those cases, so it is impossible to get `max.get<UInt8> == 0`.

But to make it more robust, I have fixed it and added some unit tests.